File: Algorithms_medium/78_Subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(u, t):
            if u == len(nums):
                print(t)
                ans.append(t[:])
                return
            dfs(u + 1, t)
            t.append(nums[u])
            dfs(u + 1, t)
            t.pop()
        
        ans = []
        dfs(0, [])
        return ans
'''
dfs深度搜索算法
要同时兼顾数字扩展和位数扩展两个方向，第一个dfs相当于扩展数字，第二个相当于扩展位数
要想通最直接把数字列出来，一看就懂
'''

File: Algorithms_C++/78_Subsets.cpp
class Solution {
public:
    vector<vector<int>> ans;
    vector<int> t;

    void dfs(int i, vector<int>& nums){
        if (i == nums.size()){
            ans.push_back(t);
            return;
        }
        t.push_back(nums[i]);
        dfs(i+1, nums);
        t.pop_back();
        dfs(i+1, nums);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        dfs(0, nums);
        return ans;
    }
};